share/scripts/inverse/iie.py

```diff
@@ -0,0 +1,1005 @@
+        ndx = np.where(array > cut_off)[0][0]
+        return ndx
+
+
+def r0_removal(*arrays):
+    r0_removed = False
+    if np.isclose(arrays[0][0], 0.0):
+        r0_removed = True
+        arrays = tuple(map(lambda a: a[1:], arrays))
+    return r0_removed, arrays
+
+
+def calc_c(r, g_tgt, G_minus_g, n, rho):
+    r0_removed, (r, g_tgt, G_minus_g) = r0_removal(r, g_tgt, G_minus_g)
+    # total correlation function h
+    h = g_tgt - 1
+    # FT of h
+    omega, h_hat = fourier(r, h)
+    # direct correlation function c from OZ
+    if n == 1:
+        # single bead case
+        c_hat = h_hat / (1 + rho * h_hat)
+    else:
+        _, G_minus_g_hat = fourier(r, G_minus_g)
+        c_hat = h_hat / ((1 + n * rho * G_minus_g_hat)**2
+                         + (1 + n * rho * G_minus_g_hat) * n * rho * h_hat)
+    _, c = fourier(omega, c_hat)
+    if r0_removed:
+        c = np.concatenate(([np.nan], c))
+    return c
+
+
+def calc_g(r, c, G_minus_g, n, rho):
+    r0_removed, (r, c, G_minus_g) = r0_removal(r, c, G_minus_g)
+    omega, c_hat = fourier(r, c)
+    if n == 1:
+        h_hat = c_hat / (1 - rho * c_hat)
+    else:
+        _, G_minus_g_hat = fourier(r, G_minus_g)
+        h_hat = ((c_hat * (1 + n * rho * G_minus_g_hat)**2)
+                 / (1 - n * rho * (1 + n * rho * G_minus_g_hat) * c_hat))
+    _, h = fourier(omega, h_hat)
+    g = h + 1
+    if r0_removed:
+        g = np.concatenate(([np.nan], g))
+    return g
+
+
+def calc_dc_ext(r_short, r_long, c_k_short, g_k_short, g_tgt_short, G_minus_g_short, n,
+                rho):
+    """Calculate Δc_ext with netwon method. Jacobian has an implicit extrapolation of c
+    with zeros on twice its original range."""
+    # _k is iteration k
+    # _s is short
+    # _tgt is target
+    r0_removed, (r_short, c_k_short, g_k_short, g_tgt_short,
+                 G_minus_g_short) = r0_removal(r_short, c_k_short, g_k_short,
+                                               g_tgt_short, G_minus_g_short)
+    F = gen_fourier_matrix(r_long, fourier)
+    Finv = np.linalg.inv(F)
+    B = np.concatenate((np.diag(np.ones(len(r_short))),
+                        np.zeros((len(r_long) - len(r_short), len(r_short)))),
+                       axis=0)
+    Binv = np.linalg.pinv(B)
+    # J = Binv @ (Finv @ np.diag((1 + n * rho * F @ B @ G_minus_g_short)**2
+    # / (1 - (1 + n * rho * F @ B @ G_minus_g_short)
+    # * n * rho * F @ B @ c_k_short)**2) @ F) @ B
+    F_B_G_minus_g_short = np.matmul(F, np.matmul(B, G_minus_g_short))
+    F_B_c_k_short = np.matmul(F, np.matmul(B, c_k_short))
+    J = np.matmul(
+        np.matmul(Binv,
+                  np.matmul(
+                      np.matmul(Finv,
+                                np.diag((1 + n * rho * F_B_G_minus_g_short)**2
+                                        / (1 - (1 + n * rho * F_B_G_minus_g_short)
+                                           * n * rho * F_B_c_k_short)**2)),
+                      F)
+                  ),
+        B)
```

Comment:
I would split this snippet into smaller components. Just for readability

Reply:
or use the `np.matrix` syntax here to get it a lot more concise.

Reply:
Good idea, will do.

I was first using @, but then python needs to be 3.5+, so I switched to matmul.

Reply:
Ah ok, I am not going to use np.matrix, it is no longer recommended. See https://numpy.org/doc/stable/reference/generated/numpy.matrix.html

Reply:
Concerning python 3.5. I think that is more than reasonable considering how current our compilers etc have to be. 

@felipeZ any thought?

Reply:
A couple of comments:
*  `matmul`, `dot` and `@` *do exactly the same thing when operating over two matrices* (Their behavior differ on other cases):
```py
xs = np.random.normal(size=10).reshape(5,2)
ys = np.random.normal(size=10).reshape(2,5)
mat1 = np.dot(xs, ys)
mat2 =np.matmul(xs, ys)
mat3 = xs @ ys

assert all((np.allclose(mat1, mat2), np.allclose(mat1, mat3), np.allclose(mat2,mat3)))
```
* About the Pyhon 3.5+ version, it is a very safe assumption because most of the Linux distributions already comply with it.

Reply:
I much prefer the `@` operator. Should there be a check for Python 3.5+?

Reply:
@junghans will make sure in the cmake I think.

Reply:
> @junghans will make sure in the cmake I think.

Done